docs/operating-scylla/nodetool-commands/restore.rst

```diff
@@ -11,27 +11,34 @@ Syntax
 
    nodetool [(-h <host> | --host <host>)] [(-p <port> | --port <port>)]
                --endpoint <endpoint> --bucket <bucket>
-               --snapshot <snapshot>
-               --keyspace <keyspace> [--table <table>]
+               --prefix <prefix>
+               --keyspace <keyspace>
+               --table <table>
                [--nowait]
+               <sstables>...
 
 Example
 -------
 
 .. code-block:: console
 
-   nodetool restore --endpoint s3.us-east-2.amazonaws.com  --bucket bucket-foo --snapshot ss --keyspace ks
+   nodetool restore --endpoint s3.us-east-2.amazonaws.com  --bucket bucket-foo --prefix ks/cf/24601 --keyspace ks --table cf \
+     scylla/ks/cf-6540e8f0ce1211ee956eb1b3dcd214e1/me-3gdq_0bki_2dy4w2gqj6hoso4mw1-big-TOC.txt \
+     scylla/ks/cf-6540e8f0ce1211ee956eb1b3dcd214e1/me-3gdq_0bki_2dipc1ysb2x2a3btgh-big-TOC.txt \
+     scylla/ks/cf-6540e8f0ce1211ee956eb1b3dcd214e1/me-3gdq_0bki_2s3e829t3gyq994yjl-big-TOC.txt
```

Comment:
Shouldn't these be just the base names of the toc files?

Reply:
@cezarmoise please refer to https://github.com/scylladb/scylladb/issues/20461 for the background. in the meanwhile, i changed the keys in the example to make it more meaningful.

Reply:
@cezarmoise do you still have concerns on this?

Reply:
My question was based around what I observed when using it.
After a backup, on the bucket I see files with paths like `{prefix}/{basename}`

For example, using the prefix `backup` I get this:
```
{'Key': 'backup/manifest.json', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 893000, tzinfo=tzutc()), 'ETag': '"f6d6d0310ff5b9e2caa17d37ff7c0df3"', 'Size': 112, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-CRC.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 884000, tzinfo=tzutc()), 'ETag': '"250ac1a14852bca52deb573cce75984e"', 'Size': 8, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Data.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 886000, tzinfo=tzutc()), 'ETag': '"4d5dbcca82f7ab49aa965dffa176d312"', 'Size': 358, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Digest.crc32', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 889000, tzinfo=tzutc()), 'ETag': '"93392082d1a3d246ee470543e45e0c05"', 'Size': 10, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Filter.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 885000, tzinfo=tzutc()), 'ETag': '"2ba870e758f8154ff845970ae5530bb7"', 'Size': 24, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Index.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 888000, tzinfo=tzutc()), 'ETag': '"c6a1bd1bdfa1facf47a49ff8381d6c75"', 'Size': 56, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Scylla.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 923000, tzinfo=tzutc()), 'ETag': '"72c76ab32ec3eb302c2aa9e4ee45a36a"', 'Size': 73075, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Statistics.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 890000, tzinfo=tzutc()), 'ETag': '"44d8d3ddd8abf90b878360d8c39c4a3a"', 'Size': 4668, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-Summary.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 887000, tzinfo=tzutc()), 'ETag': '"e9cff3ea8c6cd8b37148a79f0c7e39a1"', 'Size': 50, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-TOC.txt', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 891000, tzinfo=tzutc()), 'ETag': '"f9025aa3c4e7269fc80807754caa6ef1"', 'Size': 90, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-CRC.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 878000, tzinfo=tzutc()), 'ETag': '"d70655d5f82fdd542737c198cc273f56"', 'Size': 8, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Data.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 880000, tzinfo=tzutc()), 'ETag': '"7d72af78060f066700cf7ac0db33fe06"', 'Size': 551, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Digest.crc32', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 882000, tzinfo=tzutc()), 'ETag': '"f9f5becaee67a7f10f94915cf610e19d"', 'Size': 10, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Filter.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 880000, tzinfo=tzutc()), 'ETag': '"1dfc1dd94310d168484c817447068f19"', 'Size': 32, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Index.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 881000, tzinfo=tzutc()), 'ETag': '"e5970c909ef106c2dda6db41271dbe7e"', 'Size': 89, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Scylla.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 878000, tzinfo=tzutc()), 'ETag': '"ccc4d4c6dc197580fb1957934a9e8d03"', 'Size': 75253, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Statistics.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 882000, tzinfo=tzutc()), 'ETag': '"177233f2b4b20539e5d4a8bef5062d6e"', 'Size': 4668, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-Summary.db', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 881000, tzinfo=tzutc()), 'ETag': '"032c9be59d9e4e2f05b9e8cd11f989b0"', 'Size': 50, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/me-3gkf_0mq4_3u8uo2lq7zmitlf9k6-big-TOC.txt', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 883000, tzinfo=tzutc()), 'ETag': '"f9025aa3c4e7269fc80807754caa6ef1"', 'Size': 90, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}},
{'Key': 'backup/schema.cql', 'LastModified': datetime.datetime(2024, 10, 16, 8, 10, 52, 892000, tzinfo=tzutc()), 'ETag': '"2e5075be5be3fc4fca7fd025af2ce446"', 'Size': 641, 'StorageClass': 'STANDARD', 'Owner': {'DisplayName': 'minio', 'ID': '02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4'}}

```

and to restore I used `me-3gkf_0mq4_3u8uo24v9d22dxj292-big-TOC.txt` as the sstable param and `backup` as prefix

But if I backup with `backup/partial` as prefix, and try to restore with `backup` as prefix and `partial/me-3gkf_0mq4_3u8uo24v9d22dxj292-big-TOC.txt`, then it doesn't work (S3 object doesn't exist).

The unit tests also use just the basename.

Reply:
@cezarmoise are you able to reproduce the failure using `scylla nodetool backup` and `scylla nodetool restore`?

Reply:
Also tried in the unit tests, by adding an extra path to the prefix on backup, and same to the tocs on restore.
It fails with same error.

```diff
diff --git a/test/object_store/test_backup.py b/test/object_store/test_backup.py
index 932b4caee4..9d0506940a 100644
--- a/test/object_store/test_backup.py
+++ b/test/object_store/test_backup.py
@@ -153,10 +153,10 @@ async def test_simple_backup_and_restore(manager: ManagerClient, s3_server):
     orig_res = cql.execute(f"SELECT * FROM {ks}.{cf}")
     orig_rows = { x.name: x.value for x in orig_res }
 
-    toc_names = [entry.name for entry in list_sstables() if entry.name.endswith('TOC.txt')]
+    toc_names = [f"extra/{entry.name}" for entry in list_sstables() if entry.name.endswith('TOC.txt')]
 
     prefix = f'{cf}/{snap_name}'
-    tid = await manager.api.backup(server.ip_addr, ks, cf, snap_name, s3_server.address, s3_server.bucket_name, prefix)
+    tid = await manager.api.backup(server.ip_addr, ks, cf, snap_name, s3_server.address, s3_server.bucket_name, f'{prefix}/extra')
     status = await manager.api.wait_task(server.ip_addr, tid)
     assert (status is not None) and (status['state'] == 'done')
```

Reply:
Not a blocker. You can resolve this comment.